From ee4191acf3bef59852c8ff934bf2a9d4d6e857c3 Mon Sep 17 00:00:00 2001
From: Phil Edworthy <phil.edworthy@renesas.com>
Date: Tue, 4 Oct 2016 15:14:24 +0100
Subject: [PATCH] net: dw: Support fixed links

This allows the GMAC to be used when connected directly to a switch,
i.e. when the link is always up and the speed is fixed.

Signed-off-by: Phil Edworthy <phil.edworthy@renesas.com>
---
 drivers/net/designware.c | 25 +++++++++++++++++++++++--
 drivers/net/designware.h |  1 +
 include/netdev.h         |  1 +
 3 files changed, 25 insertions(+), 2 deletions(-)

diff --git a/drivers/net/designware.c b/drivers/net/designware.c
index e2c90dc..8305ef2 100644
--- a/drivers/net/designware.c
+++ b/drivers/net/designware.c
@@ -327,7 +327,22 @@ static int _dw_eth_init(struct dw_eth_dev *priv, u8 *enetaddr)
 		return ret;
 	}
 
-	dw_adjust_link(mac_p, priv->phydev);
+	/* Fixed link? */
+	if (priv->fixed_link_speed) {
+		u32 conf = readl(&mac_p->conf) | FRAMEBURSTENABLE | DISABLERXOWN;
+
+		if (priv->fixed_link_speed != 1000)
+			conf |= MII_PORTSELECT;
+
+		if (priv->fixed_link_speed == 100)
+			conf |= FES_100;
+
+		conf |= FULLDPLXMODE;
+
+		writel(conf, &mac_p->conf);
+	} else {
+		dw_adjust_link(mac_p, priv->phydev);
+	}
 
 	if (!priv->phydev->link)
 		return -EIO;
@@ -531,7 +546,7 @@ static int dw_write_hwaddr(struct eth_device *dev)
 	return _dw_write_hwaddr(dev->priv, dev->enetaddr);
 }
 
-int designware_initialize(ulong base_addr, u32 interface)
+int designware_initialize_fixed_link(ulong base_addr, u32 interface, int speed)
 {
 	struct eth_device *dev;
 	struct dw_eth_dev *priv;
@@ -577,12 +592,18 @@ int designware_initialize(ulong base_addr, u32 interface)
 	eth_register(dev);
 
 	priv->interface = interface;
+	priv->fixed_link_speed = speed;
 
 	dw_mdio_init(dev->name, priv->mac_regs_p);
 	priv->bus = miiphy_get_dev_by_name(dev->name);
 
 	return dw_phy_init(priv, dev);
 }
+
+int designware_initialize(ulong base_addr, u32 interface)
+{
+	return designware_initialize_fixed_link(base_addr, interface, 0);
+}
 #endif
 
 #ifdef CONFIG_DM_ETH
diff --git a/drivers/net/designware.h b/drivers/net/designware.h
index d345c5b..4b80d61 100644
--- a/drivers/net/designware.h
+++ b/drivers/net/designware.h
@@ -242,6 +242,7 @@ struct dw_eth_dev {
 
 	struct phy_device *phydev;
 	struct mii_dev *bus;
+	int fixed_link_speed;
 };
 
 #ifdef CONFIG_DM_ETH
diff --git a/include/netdev.h b/include/netdev.h
index 028672e..6edb217 100644
--- a/include/netdev.h
+++ b/include/netdev.h
@@ -35,6 +35,7 @@ int cs8900_initialize(u8 dev_num, int base_addr);
 int davinci_emac_initialize(void);
 int dc21x4x_initialize(bd_t *bis);
 int designware_initialize(ulong base_addr, u32 interface);
+int designware_initialize_fixed_link(ulong base_addr, u32 interface, int speed);
 int phy_adjust_link_notifier(struct phy_device *phydev);
 int dm9000_initialize(bd_t *bis);
 int dnet_eth_initialize(int id, void *regs, unsigned int phy_addr);
-- 
2.7.4

